keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,874 @@
+plan to reject non-canonical IPv6 addresses; that will remain a
+warning only.)
+
+All in-tree API types have already been modified to use appropriate
+`utilvalidation` methods for IP/CIDR validation ([kubernetes #122931],
+[kubernetes #123174]), so we just need to update the validation
+functions themselves to allow only unambiguous values.
+
+(UPDATE: the newly-added DRA `NetworkDeviceData` field uses
+`netutils.ParseCIDRSloppy` directly and needs to be fixed.)
+
+[kubernetes #122931]: https://github.com/kubernetes/kubernetes/pull/122931
+[kubernetes #123174]: https://github.com/kubernetes/kubernetes/pull/123174
+
+#### Ratcheting validation for pre-existing objects
+
+We want to allow making changes to objects containing invalid
+IPs/CIDRs, as long as the IPs/CIDRs themselves are not changed.
+
+In the case of the four fields above which are not arrays themselves,
+and not contained within arrays, this is simple: if `pod.spec.hostIP`,
+`pod.spec.podIP`, `node.status.podCIDR`, or `service.spec.clusterIP`
+contains an invalid value, and an update the the Pod/Service does not
+change the value of that field, then the invalid value should be
+allowed to remain.
+
+In the case of array-valued fields, this becomes more complicated; if
+a service has `clusterIPs: ["001.002.003.004"]`, should it be possible
+to update it to `clusterIPs: ["001.002.003.004", "fd00::1234"]`? Does
+that count as not changing the original invalid IP or not?
+
+The current implementation of this KEP allows such changes in the case
+of array-valued fields that are not themselves within arrays
+(`node.spec.podCIDRs[]`, `pod.spec.dnsConfig.nameservers[]`,
+`service.spec.clusterIPs[]`, `service.spec.externalIPs[]`,
+`service.spec.loadBalancerSourceRanges[]`, and
+`serviceCIDR.spec.cidrs[]`) and fields whose immediate parent is an
+array but is not within any other array (`pod.spec.hostAliases[].ip`,
+`pod.status.hostIPs[].ip`, `pod.status.podIPs[].ip`,
+`service.status.loadBalancer.ingress[].ip`, and
+`ingress.status.loadBalancer.ingress[].ip`). In these cases, any
+invalid IP that appears anywhere in the array in the old object is
+allowed to appear anywhere in the array in the updated object.
+However, newly-added IPs/CIDRs must still validate (even if there are
+also invalid pre-existing IPs/CIDRs in the array).
+
+For the case of NetworkPolicy, Endpoints, EndpointSlice, and
+NetworkDeviceData, validation is more complicated, because there are
+multiple levels of arrays above the IP/CIDR values
+(`networkpolicy.spec.egress[].to[].ipBlock.cidr`,
+`networkpolicy.spec.egress[].to[].ipBlock.except[]`,
+`networkpolicy.spec.ingress[].from[].ipBlock.cidr`,
+`networkpolicy.spec.ingress[].from[].ipBlock.except[]`,
+`endpoints.subsets[].addresses[].ip`,
+`endpoints.subsets[].notReadyAddresses[].ip`,
+`endpointslice.endpoints[].addresses[]`,
+`resourceclaim.status.devices[].networkData.ips[]`).
+
+In the current implementation, NetworkPolicy obeys an extended version
```

Comment:
Again, for declarative:  This is going to be tricky.  These fields are currently `+listType=atomic` and could not be made into `listMap`.  We will need to think about how to extract some info from oldValue and pass it down into the validators for newValue, or somehow modify the call.

@aaron-prindle @yongruilin @jpbetz FYI

Reply:
Arguably I'm being too clever here and we could just do with NetworkPolicy what I did for Endpoints/EndpointSlice and say that if you change anything, you have to fix everything.

Reply:
+1